presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -246,19 +266,51 @@ private Estimate calculateNullsFraction(Map<String, PartitionStatistics> statist
                     if (partitionsWithStatisticsCount == 0) {
                         return OptionalDouble.empty();
                     }
-                    else {
-                        return OptionalDouble.of(totalPartitionsCount / partitionsWithStatisticsCount * nullsCount);
-                    }
+                    return OptionalDouble.of(totalPartitionsCount / partitionsWithStatisticsCount * nullsCount);
                 });
 
-        if (totalNullsCount.isValueUnknown() || totalRowsCount.isValueUnknown()) {
+        if (totalNullsCount.isValueUnknown()) {
             return Estimate.unknownValue();
         }
-        if (totalRowsCount.getValue() == 0.0) {
-            return Estimate.zeroValue();
+        return new Estimate(totalNullsCount.getValue() / rowCount.getValue());
+    }
+
+    private Estimate calculateDataSize(
+            Map<String, PartitionStatistics> statisticsSample,
+            String columnName,
+            Estimate rowCount,
+            Estimate nullsFraction,
+            OptionalDouble rowsPerPartition)
+    {
+        if (rowCount.isValueUnknown() || !rowsPerPartition.isPresent()) {
```

Comment:
This is what I had in the latest version from my PR.   I think it's simpler.
```
private Estimate calculateDataSize(Map<String, PartitionStatistics> statisticsByPartitionName, String column, int numberOfPartitions)
    {
        List<Double> knownPartitionDataSizes = statisticsByPartitionName.values().stream()
                .filter(stats -> stats.getBasicStatistics().getRowCount().isPresent()
                        && stats.getColumnStatistics().containsKey(column)
                        && stats.getColumnStatistics().get(column).getAverageColumnLength().isPresent()
                        && stats.getColumnStatistics().get(column).getNullsCount().isPresent())
                .map(stats -> {
                    double averageColumnLength = stats.getColumnStatistics().get(column).getAverageColumnLength().getAsDouble();
                    long rowCount = stats.getBasicStatistics().getRowCount().getAsLong();
                    long nullsCount = stats.getColumnStatistics().get(column).getNullsCount().getAsLong();
                    long nonNullsCount = rowCount - nullsCount;
                    return averageColumnLength * nonNullsCount;
                })
                .collect(toImmutableList());

        double knownPartitionDataSizesSum = knownPartitionDataSizes.stream().mapToDouble(a -> a).sum();
        long partitionsWithStatsCount = knownPartitionDataSizes.size();

        if (partitionsWithStatsCount == 0) {
            return Estimate.unknownValue();
        }
        return new Estimate(knownPartitionDataSizesSum * numberOfPartitions / partitionsWithStatsCount);
```

Reply:
> what's the usefulness of rowsPerPartition? From my perspective it doesn't seem that useful.

this was extracted opportunistically -- this value was used in 2 existing places and i added 2 new usages

> It'll give weird results if you have very variable partitions.

There is no ultimate cure for that, is there?

> This is what I had in the latest version from my PR. I think it's simpler.

well, this PR was initially a lot simpler, but grew complicated along the way. If you don't feel strongly about this, I will keep it as it is.

Reply:
To clarify-- my question wasn't about why pass in rowsPerPartition vs. compute it.  It was why is rowsPerPartition (which means something like "average number of rows per-partition for the partitions that we know how many rows they have") needed in this method at all.  IMO it complicates the method without providing a whole lot of value. I don't think it would be very common for the value to actually be used (because generally we either won't know the averageColumnLength or we'll also know the number of rows or possibly we won't know the number of rows for any partitions).  For  the rare case where the value would get used, something weird is probably going on with the partitions, and so the estimate it provides probably isn't that good.  Please let me know if I'm misunderstanding something

I don't care about using my code vs. a version of this. That was to illustrate how some of the complexity could be trimmed.

Reply:
@arhimondr says his upcoming pr is going to change this code anyway, so I'll approve.